### DISCUSSION TITLE
How to use Vec<&str> as pyo3 struct field?

### DISCUSSION THREAD

**Discussion by tushushu** [PyO3/pyo3#2128] [Questions] [answered]
It seems that the lifetime is not supported in pyo3, so what should I do if I need declare the struct field as `Vec<&str>` type?
`Vec<String>` is not used here because:
1. There would be a lot of duplicates in the vector;
2. The elements of the vector are not mutable.
Thanks very much~

**Comment by birkenfeld** [comment, marked as answer]
What you could normally use in Rust for such cases is `Vec<Rc<str>>` which avoids the duplicates.  I don't think this can be directly be exposed to Python though.

What about `Vec<Py<PyString>>`? If you have to expose the strings to Python, you'll have to convert them to PyString at some point anyway, and `Py` is like `Rc` in that you can clone them by just making a new reference.  Python strings are also immutable.

**Reply by tushushu** [reply]
Thanks @birkenfeld , I am trying to write a small Python project to process array-like object more efficiently. So I prefer to use native rust string type here.

**Reply by mejrs** [reply]
Can you give an example of what you're trying to do?

If you are trying to have structs like this...
```rust
struct Foo<'a> {
    things: Vec<&'a str>
}
```
...then someone else will have to own whatever those `&str'`s are pointing to. As such these structs don't work very well because they'll be bound to something else. You almost always want structs to *own* their fields.

That's true in general and not specific to pyo3. But when you add python into the mix and give this struct to it, it can keep it alive forever. As such, whatever is backing the references has to remain alive forever. This is why the pyclass macro forbids lifetime parameters.

**Reply by tushushu** [reply]
Currently the implementation is as below, I am using `RefCell<Vec<String>>` instead of `RefCell<Vec<&str>>` to avoid the lifetime issue. [Full implementation](https://github.com/tushushu/ulist/blob/main/ulist/src/string.rs)

When handling things like 'reques_id' or 'uuid' it's fine because there are little duplicate values. 

But when processing things like 'gender' or 'education background', then it's a waste of memory. For there are only 3 or 4 unique elements, we'd better not creating so many `String` objects on the heap.

I am also thinking of having a `HashMap` in the struct to keep the unique strings as keys and number of this key in the vector as values, and having a Vec<Pointers> to point to the `HashMap`. But maybe this is too tricky, and there may be a simple way to solve this issue?

By the way, I am new to Rust. So if there is anything I said is inaccurate please correct me, thanks very much.

```Rust
#[pyclass]
pub struct StringList {
    _values: RefCell<Vec<String>>,
}

#[pymethods]
impl StringList {
    // Arrange the following methods in alphabetical order.

    #[new]
    pub fn new(vec: Vec<String>) -> Self {
        List::_new(vec)
    }

    pub fn append(&self, elem: String) {
        List::append(self, elem)
    }

    pub fn as_bool(&self) -> BooleanList {
        AsBooleanList::as_bool(self)
    }

    ......

    pub fn to_list(&self) -> Vec<String> {
        List::to_list(self)
    }

    pub fn union_all(&self, other: &Self) -> Self {
        List::union_all(self, other)
    }

    pub fn unique(&self) -> Self {
        NonFloatList::unique(self)
    }
}
```

**Reply by mejrs** [reply]
If you want to share the string, you probably want a `Rc<str>`.

> things like 'gender' or 'education background'

Try using enums for these rather than strings, if you can.

Also - unless I'm missing something I'm not sure if you need `RefCell`, since you're not sharing it.

**Reply by tushushu** [reply]
> If you want to share the string, you probably want a `Rc<str>`.
> 
> > things like 'gender' or 'education background'
> 
> Try using enums for these rather than strings, if you can.
> 
> Also - unless I'm missing something I'm not sure if you need `RefCell`, since you're not sharing it.

Yeah, let me try `Rc<str>` and enums. I am using `RefCell` here to get mutability because the struct field `Vec<T>` is immutable. 
Thanks very much for helping me with this. @mejrs @birkenfeld
